### ISSUE TITLE
Event content type

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa#2396]
This issue has been re-written based on recent discussions.

Building upon #3170, the idea is to extend the event content type for Admissions and create a separate content type for "Event Series".

## Proposed Solution

As a webmaster I can create an Event Series piece of content for marketing purposes that provides a general overview and pulls together related event instances without sacrificing SEO.

- [x] #4010
- [x] #4018
- [ ] Revisit other goodies with the details section.

<details>
<summary>Original issue summary</summary>

# Motivation
Admissions has expressed interest in an event content type. They prefer to not use the University Events Calendar based on past experiences and because their events shouldn't be listed on events.uiowa.edu. There is also extra overhead with adding events through content.uiowa.edu because of the approval process. They also have a technical requirement of relating counselors with events to be able to display a counselor on a specific event page.

# Additional Requirements
They would like the ability to display a subset of events on the homepage. They currently use the promoted to front page option to allow for this.

They also would like a calendar display of events. They are currently using Full Calendar to accomplish this.

There is not a need currently to display past events.

They currently have the ability but are not using repeating events functionality.

Initial Epic covered by:
- [x] Local event content type feature split https://github.com/uiowa/uiowa/issues/3170
- [ ] ~~SiteNow Events view page https://github.com/uiowa/uiowa/issues/3171~~
- [x] #3773
- [ ] Auto archive local event https://github.com/uiowa/uiowa/issues/3172
- [x] Admissions local event extended https://github.com/uiowa/uiowa/issues/3173
- [x] Establish a way to create a "series" or "schedule" of similar events either through a separate content type or event type designation. Prioritize making the UX of adding multiple event instances together easy and quick. Ability to show series information by default and override on event instances. - https://schema.org/EventSeries, https://schema.org/eventSchedule
  - Options for this "series" content could be the page content type, a separate content type or a event type on the event content form.
    - Page Pros - Flexibility and we could create a related events block that could be placed. Cons - no structured data to inherit/populate event instances. 
    - Separate content type Pros - structured and ability to set the proper SEO defaults. event instances could be an ief field. Cons - one more content type to mange for a feature that won't be turned on by default. Complexity to figure out how to pass on data to event instances.
    - Event content type Pros - keep it all in one place. Cons - swiss army knife type complexity, lots of uses, not great at any one thing.
- [x] Change the Event Schema.org dates to be ISO 8601 (https://github.com/uiowa/uiowa/pull/3483)
- [x] Use the same country trick as person contact information to provide US as default but not render it if nothing else is set.

## Additional Feature Split Functionality down the road

- Logic to automatically craft/update event schema
  - Hide on lists = no-follow?
- Website (link with optional title)
- Option to show views page tab with past
- ~~Parent Event Designation boolean~~ _Source link_
- ~~Reference Parent Designated Event~~ _Source link_
- Event status options (e.g. cancelled, rescheduled, postponed, moved online. https://schema.org/EventStatusType)
- https://schema.org/eventAttendanceMode (not official spec)
- Moved boolean
- Ticket/Registration price range
- Ticket/Registration (link with optional title)
- Policies (e.g. refunds) (minimal text area)
- Performers (paragraphs bundle collecting name, phone, email, url)
- Organizers (paragraphs bundle collecting name, phone, email, url)

## _Shooot..._ content.uiowa.edu replacement?

- JSON feeds
- Filters/categorizations
- Event admins/units
- Migrate import? Track changes? https://mglaman.dev/blog/tracking-changes-migrate-dynamic-row-hashes

</details>

**Comment by briand44** [comment]
@scottfi - I am curious about the Virtual featured event field. Is this currently being used and is this something you would like to continue to exist? If so, can you explain how this functions?

Also, there currently is the ability to upload image(s) that would then appear in a slideshow at the top of the page. Is this functionality needed or is there any need for the ability to add images outside of what is available in the WYSIWYG that will exist on the body field?

**Comment by scottfi** [comment]
The virtual and featured event fields are something we added when we moved all our visit events to virtual/zoom last spring. I think the answer to whether it is needed or not depends on when our new site rolls out (and if we're still offering virtual only  events). We've switched this content type around several times over the past 4-5 months to adjust for the types of events we've been offering.

Virtual Event: This was a flag that we were using to pull zoom events into a specific calendar that we were displaying on our visit pages as an alternative to listing links to those those events directly. We set up a view and a block to pull virtual events and also to sort by type. The virtual event flag was just a way for us to keep those events separate from normal in-person events that we previously used that content type for.

Featured Event: This flag was a way for us to add a bit of hierarchy to our virtual events. We have larger general events (Virtual Hawkeye Visit Days, Transfer Days, Iowa Overviews, etc.) that are more important than the smaller academic-specific virtual events. and this allowed us to bubble those to the top or feature them more prominently in the large list of virtual events that we offered.

Let me know if you want to talk through this quickly - I can probably show you examples of how those are used which might help explain them.

**Comment by pyrello** [comment]
@mark-bennett-uiowa Wanted to check in with you about how you intend to use local events on your sites and how that might intersect with this.

**Comment by mark-bennett-uiowa** [comment]
@pyrello I have preferred that departments and other university units in CLAS use the event system. They have the ability to post events on content hub and mark them as private if they dont want the events showing up on events.uiowa.edu. However, there is the possibility that a center (e.g. The Sloan Center) either doesnt have the ability or the tagging capabilities of doing this. So, in the rare occurrence that they cant possibly use content hub, that we could use a local event option on the site only. Because of limitations with views in D7, the ability to "interweave" local and content hub events was difficult, if not impossible, so I dont want to get into the habit of continuing that into D8 for the CLAS sites.

If a local event was available for these small one-offs, I would be for it, but my priority has been to push them towards content hub first. I hope that makes sense. :)

**Comment by bspeare** [comment]
I added this related issue https://github.com/uiowa/uiowa/issues/2727 to document the conversation we had about event listings today.

**Comment by mjoneill** [comment]
Not Admissions-specific, but this seems like the best place to drop in some additional thoughts I'm hearing from across campus about Events...  Of particular note is the handling of expired events (some campus partners have a need to keep those listed).

Notes on Events

Recent conversations with stakeholders for next-gen arts.uiowa.edu have surfaced some interesting discussions around the events calendar that might be of interest as you think about work on events.

### Deleting expired events
Users report that once an event has passed its scheduled date/time, they are removed.  This creates problems for Music and others in the Arts community (and presumably anyone who runs into this use case) because for many of their events they have the “live performance date”, but want the event to stay listed for X weeks afterwards (where X was suggested to be 2 weeks during the conversation) because it is available after the fact as a recording. Stakeholders report having to resort to various stratagems to work around this “delete old events” functionality, including creating multiple events, one for each day of the X week duration, to “simulate” the functionality they want.    From their perspective, it would be much more preferable for expired events to stay listed and visible in chronological order.

Stakeholders did not seem willing to commit to a universal change, whereby all events would live forever…but instead seemed interested in some sort of optional checkbox/control that allowed them to select events that should persist for posterity purposes.

Note: There are some problems <https://www.searchenginejournal.com/the-essential-guide-to-managing-expired-content/318417/> with keeping every event in perpetuity, but if that’s part of the strategy, even for some events, we should look at using ValidThrough schema or an unavailable_after tag.  

 

### Long duration events
Arts stakeholders expressed frustration with how to use the events system to best represent long-duration events.  For example, Stanley Art Museum has a three month long traveling exhibit.  How should they represent that in Events?

Within that exhibit, there may even be more granular/discrete events.  For example, an event that features an artist discussion, or a lecture series, etc…. 

Note: some of these relationships between sub-events for a long duration super-event like a 3 month exhibit could be clarified to search engines with Event schema <https://schema.org/Event> using the superEvent and subEvent elements.  Maybe it helps to understand the schema that represents these types of things?


### Possibly related
Google has some guidance around postponed, cancelled, and virtual events: https://developers.google.com/search/blog/2020/03/new-properties-virtual-or-canceled-events.   Other events info: https://developers.google.com/search/docs/data-types/event , including guidelines for multi-day events

**Comment by scottfi** [comment]
I'll add my 2 cents to this specifically for our needs on the admissions site. We've actually talked about this very issue over the past several years, and do not need any sort of archival event retention. In fact, we'd like to get rid of that since our site doesn't expire events currently. We've looked into this on our site to try and retire/remove events X days after they happen for quite some time. We don't want/need these to show up in search engines after they are done since our events vary year to year and the details are different enough that an old event isn't very helpful living out there somewhere. So I think some sort of function that retires events after a certain amount of time is the direction we'd like to go with the admissions site.

**Comment by joewhitsitt** [comment]
Took a stab at grooming this issue to the point where I wanted to do a research spike into smart date. Packaged up what I discovered in #2765

**Comment by joewhitsitt** [comment]
Going to split this epic out

**Comment by joewhitsitt** [comment]
I updated the IS with a few thoughts around the "landing page" event or as I have it, "Series." Admissions said that these would be unlisted on their site (robots no follow). I think with each of the options, that should be an override and not a default.

**Comment by joewhitsitt** [comment]
> I think multiple dates/repeating is a nice feature for editors but goes against the future of the web from an SEO perspective. An event is a single instance that can be part of a series or "schedule."
> https://developers.google.com/search/docs/data-types/event
> https://schema.org/eventSchedule
> 
> > For MVP, i am thinking of moving it back to single date, no repeat functionality. That still leaves it open to handle a parent/child or series/event relationship. The epic does mention a parent designation and also mentions that Admissions has repeat functionality current but does not use it. #2396

https://github.com/uiowa/uiowa/issues/3170#issuecomment-832806823

Hearing the discussion this morning in the Admissions meeting, I think the Admissions and the 80% would be better served by an event content type that can handle multiple dates/repeat/recurring.

**Comment by joewhitsitt** [comment]
> > I think multiple dates/repeating is a nice feature for editors but goes against the future of the web from an SEO perspective. An event is a single instance that can be part of a series or "schedule."
> > https://developers.google.com/search/docs/data-types/event
> > https://schema.org/eventSchedule
> > > For MVP, i am thinking of moving it back to single date, no repeat functionality. That still leaves it open to handle a parent/child or series/event relationship. The epic does mention a parent designation and also mentions that Admissions has repeat functionality current but does not use it. #2396
> 
> [#3170 (comment)](https://github.com/uiowa/uiowa/issues/3170#issuecomment-832806823)
> 
> Hearing the discussion this morning in the Admissions meeting, I think the Admissions and the 80% would be better served by an event content type that can handle multiple dates/repeat/recurring.

The team discussed this today and from what I heard, the event date field should remain single for now. This is limited use functionality and we should strive to establish the structure that the web has defined. To solve for admissions only this field config could be switched and/or we can implement other features to satisfy requirements:

- event schedule text (long) field (to summarize long series) that feeds https://schema.org/eventSchedule
- reference other instances either through a parent designation that the children reference or referencing children separately through the parent.

**Comment by pyrello** [comment]
I'm thinking about a design for this where we have a single event content type that can be created. If all you need is a single event, you need to do nothing more than fill it out. Additionally, it could have a reference field to another entity (fragment or possibly content) that represents an event schedule, which is optional, but could be used to represent the wider world of repeating events or bespoke schedules. We could then represent the event as a single Event if the schedule is not used and represent it as an EventSchedule if it is used.

**Comment by joewhitsitt** [comment]
Oh snap... Could be good or risky. https://www.drupal.org/project/date_content allows unique content to be associated with a date instance.

**Comment by briand44** [comment]
Capturing some notes from my last meeting with Admissions.

Example of Hawkeye Visit - Occurs multiple days - Oct 8th, 11th, etc.

- Show each date on event calendar
- Each item on event calendar would link to same place (One Hawkeye Visit day page)

Would like this solution to be simpler for staff to add events than the current way.

Could work around this by structuring event landing page custom to Admissions, if needed.

**Comment by joewhitsitt** [comment]
Made changes to the IS based on conversations with Brian before he left. He has not see what I came up with so it isn't an approval

**Comment by cory-skeers** [comment]
Capturing some details from grooming:

https://schema.org/eventSchedule, https://schema.org/EventSeries <- two similar, but slightly different options.

Google appears to prefer Event Series for all individual events.

One thought is that an event schedule is a type of recurring event that you would expect to attend every instance, whereas event series is set of events which a person may attend any one (or more than one), but not necessarily all. In the former case, you would have a single event with a schedule property, and in the latter each unique time would have a unique event associated with it, as well as an event series that features info on the overall collection.

Re: individual nodes, there is the question on whether each node needs to be created individually, or if there is a way to programmatically generate a collection of events given a set of dates or recurrence rule (say, every Mon/Wed/Fri for three months...auto-generate instances for each date).

**Comment by pyrello** [comment]
I think we should close this issue in favor of the separate issues that have been opened as follow-up to the work that was originally done related to this.

**Comment by joewhitsitt** [comment]
I talked to Brian about this earlier. We need another round of grooming on this before we can close it. We should discuss the non-issue items in this epic, "Additional Feature Split Functionality down the road" and "Shooot... content.uiowa.edu replacement?" to see if those are things we want to carve off before closing this issue.

**Comment by benlewzz** [comment]
Echoing what @mark-bennett-uiowa is saying, in almost all cases, we work with our content editors to add events to Content Hub. One recent case was when a partner wanted to upload ICS files to a website, we directed them to Content Hub, because:

- Puts the event into the most visible promotion spaces (Events Calendar, After Class, Arts Iowa, Iowa Now)
- Automatically includes the accessibility/accommodations statement
- Forces event organizer to think through details in a structured way
- Reduces duplicate event postings (if they go in one place) 
- Generates ICS files (even though they are mostly broken for virtual events)

If we use local events, it's because of one of these: 

a) they don't want to promote the event
b) it's not open to the public
c) they want the event to persist in an archival manner afterwards

Just adding in case it is helpful.

**Comment by joewhitsitt** [comment]
Just want to capture this. Got an eventbrite email not too long ago.

For those unfamiliar with Eventbrite, you can create a single event or an event with recurring dates, but note, the recurring dates don't have separate URLs and "Event details and ticket types will apply to all instances."

But the email highlighted a new feature called "Collections" which allows you to reference individual events/event with recurring dates into a separate landing page and features the ability to capture a separate title, organizer, image and summary. Very similar to the event series content type we developed for Admissions but the reference and therefore workflow are reversed. Since the collection doesn't collect a date, I am not sure if they show in the same feeds as normal events.

I don't know if this is validation of our work or what, but I would say what we are building is not too far off from one of the web's biggest event platforms.

**Comment by pyrello** [comment]
It seems like we could probably close this issue. Looking at the original summary, the only thing left undone was captured in #3172. There was also recent work to improve the local event feature in #5869. At this point, this issue seems like a zombie.

**Comment by joewhitsitt** [comment]
>Revisit other goodies with the details section.

Bunch of good ideas in there that should be sliced out if they are still worth pursuing. I know that @mark-bennett-uiowa had interest in the event status stuff. And that more of this might be worth capturing in #5374

**Comment by joewhitsitt** [comment]
Also, I don't think all of the ideas from the comments have been captured

**Comment by mark-bennett-uiowa** [comment]
Added a few updates via #5939 and #5940 as far as potential additions. Mentioned these in #2936 and #5886

**Comment by bspeare** [comment]
@joewhitsitt and @briand44 do you want to keep this issue open?

**Comment by briand44** [comment]
I am ok with it closing
